pkg/spi/spi.go

```diff
@@ -16,49 +16,126 @@
-	return gitfile.Default().GetFileContents(ctx, namespace, repoUrl, filepath, ref, callback)
+func (s SPIClient) GetFileContents(ctx context.Context, name string, component v1alpha1.Component, repoUrl string, filepath string, ref string) (io.ReadCloser, error) {
+	log := ctrl.LoggerFrom(ctx)
+	spiFCRLookupKey := types.NamespacedName{Name: SPIFCR_prefix + name, Namespace: component.Namespace} //We'll use a unique component name and filepath to construct the name
+	spiFCR := &spiapi.SPIFileContentRequest{}
+	log.Info("Looking up SPIFileContentRequest CR", "name", spiFCRLookupKey.Name, "namespace", spiFCRLookupKey.Namespace)
+	err := s.K8sClient.Get(ctx, spiFCRLookupKey, spiFCR)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			spiFCR.Name = spiFCRLookupKey.Name
+			spiFCR.Namespace = spiFCRLookupKey.Namespace
+			spiFCR.Spec.RepoUrl = repoUrl
+			spiFCR.Spec.FilePath = filepath
+			spiFCR.Spec.Ref = ref
+			//add an owner reference
+			ownerReference := metav1.OwnerReference{
+				APIVersion: component.APIVersion,
+				Kind:       component.Kind,
+				Name:       component.Name,
+				UID:        component.UID,
+			}
+			spiFCR.SetOwnerReferences(append(spiFCR.GetOwnerReferences(), ownerReference))
+
+			log.Info("SPIFileContentRequest CR not found, creating a new request", "name", spiFCR.Name, "namespace", spiFCR.Namespace)
+			err = s.K8sClient.Create(ctx, spiFCR)
+			if err != nil {
+				log.Error(err, "Unable to create an SPIFileContentRequest CR", "name", spiFCR.Name, "namespace", spiFCR.Namespace)
+				return nil, &SPIFileContentRequestError{fmt.Sprintf("Failed to create an SPIFileContentRequest CR: %s", err.Error())}
+			}
+		}
+	}
+
+	return getFileContentFromSPIFCR(*spiFCR, log)
 }
 
-func DownloadDevfileUsingSPI(s SPI, ctx context.Context, namespace string, repoURL string, ref string, path string) ([]byte, error) {
-	validDevfileLocations := []string{cdqanalysis.Devfile, cdqanalysis.HiddenDevfile, cdqanalysis.HiddenDirDevfile, cdqanalysis.HiddenDirHiddenDevfile}
+func getFileContentFromSPIFCR(fcr spiapi.SPIFileContentRequest, log logr.Logger) (io.ReadCloser, error) {
+
+	if fcr.Status.Phase == spiapi.SPIFileContentRequestPhaseDelivered {
+		fcrName := fcr.Name
+		//get contents and decode them
+		fileContent := fcr.Status.Content
+		decodedFileContent, err := base64.StdEncoding.DecodeString(fileContent)
+		if err != nil {
+			return nil, &SPIFileContentRequestError{fmt.Sprintf("Failed to decode file content %s ", err)}
+		}
 
-	for _, filename := range validDevfileLocations {
-		devfileBytes, err := DownloadFileUsingSPI(s, ctx, namespace, repoURL, ref, filepath.Join("/", path, filename))
+		log.Info("SPIFileContentRequest: Decoded file contents successfully ", "name", fcrName, "namespace", fcr.Namespace)
+		return io.NopCloser(bytes.NewBuffer(decodedFileContent)), nil
+	} else {
+		log.Info("SPI Phase delivered not reached")
+		return nil, &SPIFileContentRequestError{SPIFCR_waiting_for_delivered_phase}
+	}
```

Comment:
so if i understand this correctly; if its not delivered, requeue, fetch and check status again..

Reply:
yes, we have to wait a bit